gammapy/stats/feldman_cousins.py

```diff
@@ -0,0 +1,442 @@
+                # Upper limit is the first point where this condition is true
+                if upper_limit[-1] == -1:
+                    upper_limit[-1] = x_value
+                # Lower limit is the first point after this condition is not true
+                if x == number_bins_x - 1:
+                    lower_limit[-1] = x_value
+                else:
+                    lower_limit[-1] = x_bins[x + 1]
+
+    return upper_limit, lower_limit, x_values
+
+
+def fc_fix_upper_and_lower_limit(upper_limit, lower_limit):
+    """Push limits outwards as described in the FC paper.
+
+    Parameters
+    ----------
+    upper_limit : array-like
+        Feldman Cousins upper limit x-coordinates
+    lower_limit : array-like
+        Feldman Cousins lower limit x-coordinates
+
+    Returns
+    -------
+    upper_limit : array-like
+        Feldman Cousins upper limit x-coordinates (fixed)
+    lower_limit : array-like
+        Feldman Cousins lower limit x-coordinates (fixed)
+    """
+
+    all_fixed = False
+
+    while not all_fixed:
+        all_fixed = True
+        for j in range(1,len(upper_limit)):
+            if upper_limit[j] < upper_limit[j-1]:
+                upper_limit[j-1] = upper_limit[j]
+                all_fixed = False
+        for j in range(1,len(lower_limit)):
+            if lower_limit[j] < lower_limit[j-1]:
+                lower_limit[j] = lower_limit[j-1]
+                all_fixed = False
+
+
+def fc_find_limit(x_value, x_values_input, y_values_input, do_upper_edge = True):
+    """Find the limit for a given x value.
+
+    Parameters
+    ----------
+    x_value : double
+        The measured x value for which the upper limit is wanted.
+    x_values_input : array-like
+        The x coordinates of the confidence belt.
+    y_values_input : array-like
+        The y coordinates of the confidence belt.
+    do_upper_edge : bool
+        If x_value lies on a bin border, use the upper edge of the belt.
+
+    Returns
+    -------
+    limit : double
+        The Feldman Cousins limit
+    """
+
+    limit = 0
+
+    if do_upper_edge:
+        previous_x = np.nan
+        next_value = False
+        x_values = x_values_input
+        y_values = y_values_input
+        for i in range(len(x_values)):
+            current_x = x_values[i]
+            # If the x_value did lie on the bin border, loop until the x value
+            # is changing and take the last point (that is the highest point in
+            # case points lie on top of each other.
+            if next_value == True and current_x != previous_x:
+                limit = y_values[i-1]
+                break
+            if x_value <= current_x:
+                # If the x_value does not lie on the bin border, this should be
+                # the upper limit
+                if x_value != current_x:
+                    limit = y_values[i]
+                    break
+                next_value = True
+            previous_x = current_x
+    else:
```

Comment:
I had to look at this code again and I honestly don't remember why I coded it that way. I opted for removing the whole else clause, because I no longer understand why I thought it was useful.